lib/excon/socket.rb

```diff
@@ -208,6 +224,8 @@ def read_nonblock(max_length)
         # read until EOFError, so return everything
         @read_buffer.slice!(0, @read_buffer.length)
```

Comment:
I know this already was here, but should we simply return the `read_buffer` itself here? Slicing it's entire contents would be an allocation otherwise, which doesn't seem ideal (though that could protect from mutating it, I suppose).

Reply:
Yeah, this is just a shortcut for

```
ret, @read_buffer = @read_buffer, String.new
return ret
```

I'm not sure I find it clear either!

Reply:
Yeah, similar to the comment above, in some ways there are two challenges. One is fixing some of the performance and memory characteristics, and one is trying to make things clearer and easier to understand. It's not the worst thing to understand, but it definitely takes some time for me whenever I revisit to recall how it all fits together. It may be worth spending some time on making it clearer and easier to follow, but I don't want to block these changes on that. I think it will be easier to address clarity as a separate effort than to try and mix the two together.